File: examples/parse.py
"""Developer Kit v0.1.0: print a bounded, deterministic SROW specimen summary."""

from pathlib import Path
import json

from machine_edition_devkit.parse import MachineEdition


def main() -> None:
    package = Path(__file__).resolve().parents[1] / "specimen" / "srow" / "package"
    edition = MachineEdition.load(package, validate=True)
    units = sorted(edition.units(), key=lambda unit: unit.id)[:3]
    result = {
        "package_id": edition.manifest["package_id"],
        "version": edition.manifest["version"],
        "meaning_units": [
            {"id": unit.id, "provenance_id": unit.provenance_id}
            for unit in units
        ],
    }
    print(json.dumps(result, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: examples/parse.ts
/** Developer Kit v0.1.0: bounded TypeScript parser for the public SROW specimen.
 * Run with Node.js 22.18+ after validating the package.
 * Code: MIT. Specimen records: CC BY 4.0 within the authorized public scope.
 */

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const packageDir = join(dirname(fileURLToPath(import.meta.url)), "../specimen/srow/package");
const manifest = JSON.parse(readFileSync(join(packageDir, "manifest.json"), "utf8"));
const provenance = new Set(
  readFileSync(join(packageDir, "provenance.jsonl"), "utf8")
    .split(/\r?\n/)
    .filter(Boolean)
    .map((line) => JSON.parse(line).id)
);
const units = readFileSync(join(packageDir, "meaning-units.jsonl"), "utf8")
  .split(/\r?\n/)
  .filter(Boolean)
  .map((line) => JSON.parse(line))
  .sort((a, b) => a.id.localeCompare(b.id))
  .slice(0, 3)
  .map((unit) => {
    if (!provenance.has(unit.provenance_id)) {
      throw new Error(`Unresolved provenance_id for ${unit.id}`);
    }
    return { id: unit.id, provenance_id: unit.provenance_id };
  });

console.log(JSON.stringify({
  package_id: manifest.package_id,
  version: manifest.version,
  meaning_units: units
}, null, 2));
